api/app.py

```diff
@@ -11,31 +11,176 @@
 from models.ingredient import Ingredient
 from models.association import recipe_category
 
+import re
+
+ingredient_regex = re.compile(r'^(?P<quantity>\d+(?:\.\d+)?)\s*(?P<unit>[a-zA-Z]*)\s+(?P<name>.+)$')
+
 app = Flask(__name__)
 app.config.from_object(Config)
 db.init_app(app)
 
+
 @app.route('/')
 def hello_world():
     return 'Welcome to my snack app!'
 
+
 @app.route('/api/recipes', methods=['GET'])
 def get_recipes():
     recipes = []
     for recipe in db.session.query(Recipe).all():
         recipes.append(recipe.as_dict())
     return jsonify(recipes)
 
+
+@app.route('/api/categories', methods=['GET'])
+def get_categories():
+    categories = []
+    for category in db.session.query(Category).all():
+        categories.append(category.as_dict())
+    return jsonify(categories)
+
+
 @app.route('/api/recipes/<int:recipe_id>', methods=['GET'])
 def get_recipe(recipe_id):
-    for recipe in db.session.query(Recipe).all():
-        if recipe.id == recipe_id:
-            return jsonify(recipe.as_dict())
-    return jsonify({'error': 'Recipe not found'}), 404
+    recipe = db.session.query(Recipe).get(recipe_id)
+    if not recipe:
+        return jsonify({'error': 'Recipe not found'}), 404
+    return jsonify(recipe.as_dict())
+
+
+@app.route('/api/recipes/', methods=['POST'])
+def create_recipe():
+    try:
+        name = request.json.get('name')
+        duration = request.json.get('duration')
+        pictures = ','.join(request.json.get('pictures'))
+        instructions = request.json.get('instructions')
+        categories = request.json.get('categories')
+        ingredients = request.json.get('ingredients')
+
+        if not name or not duration or not pictures or not instructions or not categories or not ingredients:
+            return jsonify({'error': 'Missing required fields'}), 400
+
+        recipe = Recipe(
+            name=name,
+            duration=duration,
+            pictures=pictures,
+            instructions=instructions
+        )
+
+        for cat in categories:
+            cat_name = cat.get('name')
+            if not cat_name:
+                return jsonify({'error': 'Missing category name'}), 400
+            category = db.session.query(Category).filter_by(name=cat_name).first()
+            if not category:
+                db.session.rollback()
+                return jsonify({'error': f'Category {cat_name} not found'}), 404
+            recipe.categories.append(category)
+
+        db.session.add(recipe)
+        db.session.flush()
+
+        ingredients_list = [item.strip() for item in ingredients[0].split(',')]
+
+        for ingr in ingredients_list:
```

Comment:
We are going to receive both `categories` and `ingredients` as lists of dicts from frontend, so they'll have the format:
```
"categories": 
  [
    {
      "name": "Pie",
      "color": "#FFD700"
    },
    {
      "name": "Italian",
      "color": "#FF0000"
    }
  ]
```
```
"ingredients": [
      {
        "name": "sugar",
        "unit": "g",
        "quantity": 150
      },
      {
        "name": "eggs",
        "unit": null,
        "quantity": 3
      }
    ]
```
We don't need to separate the `ingredients`, they are already separated into the 3 values that we need (`name`, `unit`, and `quantity`).